### ISSUE TITLE
Data : Instance mask generation

### ISSUE THREAD

**Issue by Kunaldargan** [VisualComputingInstitute/TrackR-CNN#75]
Hi, 
I have created a dummy dataset of the KITTI MOTS format with different classes verified with MOTS Tools. Kindly resolve these following doubts.

1.  I could not find a description of how to generate instance segmentation masks from Kitti format annotations.

2. Does it affect the accuracy if jpeg images are used?

3. Do overlapping segmentation masks affect the accuracy?

**Comment by pvoigtlaender** [comment]
Hi,

1. do you mean how to convert the txt format to the png format? Usually we just work with the txt format, but I think there's a conversion script for this in the mots_tools repository.

2. I don't expect any problems with jpeg

3. actually for MOTS we do not really allow overlapping masks, so you should better resolve this somehow

**Comment by Kunaldargan** [comment]
Hi, Thanks for the reply
1. Can you please point out where in code do you use txt format input because what I see if I use
python main.py configs/conv3d_sep2.

{
  # Engine settings
  "model": "conv3d_sep2",
  "task": "train_no_val",
  "dataset": "KITTI_segtrack",
  "log_verbosity": 5,
  "gpus": 1,
  #"own_dataset_per_gpu": true,
  "use_summaries": false,
  "write_summaries": false,
  "collect_run_metadata": false,

  # MaskRCNN on/off
  "add_masks": true,

  # Pretrained model from tensorpack
  "load_init": "/globalwork/krause/pretrained_models/tensorpack/PReMVOS-COCO-Mapillary-PreProcForKITTISegtrack/converted",
  # Freeze applies to the whole model, not just our backend, but that's fine since only the backend uses batchnorm
  "freeze_batchnorm": true,
  "max_saves_to_keep": 1,

  # Training settings
  "batch_size": 8,
  "learning_rates": "{1: 0.0000005}",
  "optimizer": "adam",
  "num_epochs": 5,
  "max_saves_to_keep": 1,

  # Dataset options
  "KITTI_segtrack_data_dir": "/globalwork/krause/data/KITTI_MOTS/train/",
  "MOTS_segtrack_data_dir": "/globalwork/krause/data/MOTS_challenge/train/",
  "optical_flow_path": "/globalwork/krause/data/KITTI_flow_pwc/",
  "prefer_gt_to_ignore": true,
  "use_ioa_for_ignore": true,
  "use_masks_for_ignore": true,
  "resize_mode_train": "fixed_size",
  "input_size_train": [309, 1024],
  "resize_mode_val": "fixed_size",
  "input_size_val": [375, 1242],

  "augmentors_train": ["flip", "gamma"],
  "num_parallel_calls": 6,
  "prefetch_buffer_size": 8,

  "mask_disjoint_strategy": "score",
"tracker": "hungarian", "tracker_reid_comp": "euclidean", "detection_confidence_threshold_car": 0.8469800990815324, "reid_weight_car": 1.0, "mask_iou_weight_car": 0.0, "bbox_center_weight_car": 0.0, "bbox_iou_weight_car": 0.0, "association_threshold_car": 0.8165986526897969, "keep_alive_car": 4, "reid_euclidean_offset_car": 8.810218833503743, "reid_euclidean_scale_car": 1.0090931467228708,
"detection_confidence_threshold_pedestrian": 0.9368820089063415, "reid_weight_pedestrian": 1.0, "mask_iou_weight_pedestrian": 0.0, "bbox_center_weight_pedestrian": 0.0, "bbox_iou_weight_pedestrian": 0.0, "association_threshold_pedestrian": 0.47985540892434836, "keep_alive_pedestrian": 6, "reid_euclidean_offset_pedestrian": 9.447084376750222, "reid_euclidean_scale_pedestrian": 1.3437965549876354,

  "network": {
    "resnetconv4": {"class": "ResNet101Conv4"},
    "conv3d_1": {"class": "SepConv3DOverBatch", "activation": "relu", "n_features": 1024, "init_type": "identity", "from": ["resnetconv4"], "old_order": true},
    "conv3d_2": {"class": "SepConv3DOverBatch", "activation": "relu", "n_features": 1024, "init_type": "identity", "from": ["conv3d_1"], "old_order": true},
    "frcnn": {"class": "FasterRCNN", "fastrcnn_batch_per_img": 64, "reid_dimension": 128, "reid_loss_per_class": true,
              "reid_loss_factor": 1.0, "reid_loss_variant": 1, "reid_measure": "euclidean", "from": ["conv3d_2"],
              "class_agnostic_box_and_mask_heads": true}
  }
}

Here I don't have a path to annotations files rather I give path to  "KITTI_segtrack_data_dir": "/globalwork/krause/data/KITTI_MOTS/train/", - Which is a folder containing images and instances only.

2. Next logical question is how is tracking performed during training time using only the instance masks and not the txt format annotations.
3. Now I have different object types than Pedestrians and Cars, whereas I have 3 object types of interest do I need to make necessary changes to code, I'm thinking of making a fork and making it more accessible and general.
4. Lastly, I think there's no such script in mots_tools which can convert txt to img for masks. (I have figured it out through another way but the kind of scripts used by you during training might be helpful)
5. I understand overlapping masks must not be overlapping but surely that's not a deadlock right ? or there's no way overlapping masks can be incorporated?
6. What are the use of training.txt and validation.txt in TrackR-CNN/datasets/KITTI/segtrack/training.txt


Thank you.

**Comment by pvoigtlaender** [comment]
Hi,

1. I remembered incorrently. Actually, in the code we use the png input format, not the txt format.
2. the png and the txt formats are equivalent, since the png also contains the instance ids. So we can also train the association head
3. In the code the number of classes are specified, have a look at https://github.com/VisualComputingInstitute/TrackR-CNN/blob/master/datasets/KITTI/segtrack/KITTI_segtrack.py#L20  Note that we use the pre-trained weights from mapillary and with other/more classes, you might have a mismatch in the initialization here (but it should still work if you have enough training data)
4. you're right, we only provided the other way round, i.e. png to txt format (https://github.com/VisualComputingInstitute/mots_tools/blob/master/mots_common/images_to_txt.py). I don't have a direct conversion script to png, only a script which directly created the pngs from our internal annotation format. I could share it, but it seems you already figured that out through another way
5. At least for evaluation purposes, a method has to produce disjoint masks (that's how we defined the task). The ground truth used for evaluation also has to be disjoint, otherwise the matching between ground truth masks and result masks might be ambiguous. Maybe you can just use a simple heuristic to make the masks non-overlapping, e.g. put the smaller masks on top of the larger ones. During training I think you might get away with overlapping masks.
6. I think we used these only to train a frame-wise instance segmentation model, this should probably not be important for you

**Comment by ahnonay** [comment]
Closing this, please reopen if needed

**Comment by bat3a** [comment]
can you please say where to change the inputs from png to txt

**Comment by ahnonay** [comment]
Please create a separate issue with a detailed problem description
